walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
+        break;
+      }
+
+      // No times found yet.
+      if(potentialTime.size() == 0) { 
+        // See if there is a slot before the first booked event and if this slot.
+        // is large enough for the request. 
+        if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+          // If request duration is within slot between start of day and first event (curr).
+          if((TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start()){
+            potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+          }
+        }
+        // If meeting request can fit in between end of event and end of day.
+        if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
+            potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));      
+        }
+      } else {
+        // Time slots have been found.
+        // Remove the last timerange from list (time is prevEvent.end to endOfDay) to check conflict with currEvent.
+        potentialTime.remove(potentialTime.size()-1);
+        
+        // Check to see if prev and curr events OVERLAP.
+        // Add timerange from the end of the later timerange.
+        if(prevEvent.getWhen().overlaps(currEvent.getWhen())){
+          int laterEnd = currEvent.getWhen().end();
+          if(prevEvent.getWhen().end() > laterEnd) {
+            laterEnd = prevEvent.getWhen().end();
+          }
+          potentialTime.add(TimeRange.fromStartEnd(laterEnd, TimeRange.END_OF_DAY,true));
+        } else {
+          // check GAP (...) in-between curr and prev.
+          // |--prev--|...|--curr--|
+          long gapDuration = currEvent.getWhen().start() - prevEvent.getWhen().end();
+          if(gapDuration >= request.getDuration()){
+            potentialTime.add(TimeRange.fromStartEnd(prevEvent.getWhen().end(), currEvent.getWhen().start(), false));
+          }
+          // Check there there is a gap between the end of the event and end of day large enough for request.
+          if(currEvent.getWhen().end() != TimeRange.END_OF_DAY){
+            if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
+              potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));  
+            } 
+          }
+        }
+      }
+    }
+    return potentialTime;
+  }
+  /**
+   * Finds all events that mandatory and optional attendees are booked for - removes duplicates.
+   * @param mandatory - events that mandatory attendees in the request have to attend.
+   * @param optional - events that optional attendees in the request have to attend.
+   * @return list of events.
+   */
+  public List<Event> getAllUniqueEvents(List<Event> mandatory, List<Event> optional) {
+
+    List<Event> allEvents = new ArrayList<>(mandatory);
+
+    for(Event opt : optional){
```

Comment:
Nit:  What's the difference between an 'optional' and an 'opt'? The naming makes it hard to understand what's going on here.

The reason you chose 'opt' is because you've named the inputs mandatory and optional, which hides the fact that they are lists. Best to use plural in this case, like you did with 'allEvents'. If you rename the input arguments to 'mandatoryEvents' and 'optionalEvents' then 'opt' can become 'optionalEvent'.

Reply:
Done